### PR TITLE
insights: `seriesCount` field does not fail whole gql query

### LINKED ISSUE #38951: insights: any error in the insightViews query fails all results
When loading a list of insight views using the GraphQL query `insightViews`, any error on any insight will fail the entire response and will not return data. Currently this breaks the entire product quite badly such that the insights page cannot load anything at all. Unfortunately, this state occurs because some data on the insight definition is invalid, so there is no way for the user to fix the problem.

image

We need to address a few things:
1. If any insight throws an error we should limit the affect of this error to just the single insight
2. We need to allow users to fix problems in metadata (missing contexts, etc), which means we need to return data and let them edit the insight.

### PR DESCRIPTION
**Pull request by leonore**
tackles part of #38951

* `SeriesCount` return value is made nullable
* use `errorPolicy: ’all'` on `insightViews` queries in the webapp

nb. this isn't a catch-all fix, but this creates a blueprint if we are to add more fields to query in the future. I believe that right now `SeriesCount` is the only query that can fail in this way

#### some background:
graphql allows to return both data and errors in its response. however:
* in graphql, if there is an error on a non-nullable field then its whole parent is errored ([spec](https://spec.graphql.org/October2021/#sec-Handling-Field-Errors))

so if we make a field nullable we'll be able to get both data and errors. however that must still be handled correctly by the client.
* the apollo client allows to set an error policy we can use ([docs](https://www.apollographql.com/docs/react/data/queries/#errorpolicy))

## Test plan
The following test plan was ran on both `main` and commit c8524fee4662257ab218a7c08873c77dd613754a (parent of [this PR](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/commit/b18b87a73b592b15fb1f86a3effc6cef14ecfb11) which removed `seriesCount` and the issue)

1. Test ’All insights’ page works, even when an individual insight errors
2. Test a specific dashboard page works, even when an individual insight errors
3. Test single-insight page works, even when the insight errors.

Screenshots:

### before:

| all insights | dashboard | single insight |
| ----- | ----- | ---- |
| Screenshot 2022-07-27 at 12 32 05 | Screenshot 2022-07-27 at 12 32 01 | Screenshot 2022-07-27 at 12 24 07 |

### after:

| all insights | dashboard | single insight |
| ----- | ----- | ---- |
| Screenshot 2022-07-27 at 12 29 04 | Screenshot 2022-07-27 at 12 29 09 | Screenshot 2022-07-27 at 12 28 53 |

### CODE REVIEW

**Review by chwarwick — CHANGES_REQUESTED** [review]
While this solves the specific error we ran into last time this doesn't close https://github.com/sourcegraph/sourcegraph/issues/38951 because all the other fields in the InsightView aside from Dashboards are non-null so an error with any of them would null out the entire list of InsightViews.

To address https://github.com/sourcegraph/sourcegraph/issues/38951  I think InsightViewConnection needs updating. Moving from `nodes: [InsightView!]!` to `nodes: [InsightView]!` that would mean that an error in any 1 InsightView wouldn't remove the entire list.

**Comment by leonore** [comment]
> To address #38951 I think InsightViewConnection needs updating. Moving from `nodes: [InsightView!]!` to `nodes: [InsightView]!` that would mean that an error in any 1 InsightView wouldn't remove the entire list.

yep I was already working on this

**Comment by felixfbecker** [comment]
Could the error cases (when a field would return `null` and an error) be explained in the docs of the field in the schema?

**Comment by leonore** [comment]
> Could the error cases (when a field would return `null` and an error) be explained in the docs of the field in the schema?

@felixfbecker I'm confused whether you're suggesting this regardless of the fix, or as part of the fix. could you clarify your intentions with documenting the error cases?

**Comment by leonore** [comment]
I had a sync with @vovakulikov about the frontend implications of this change and I don't believe making `InsightViewNode` nullable in the `insightViews` query is sustainable
* it requires changes on the frontend such that we always need to handle a potential null node
* it means we need to build a fake insight object with the data we *do* have, so that it is displayed to the user with the correct message and also editable if needed to fix (e.g. search context)

an alternative would have been to return the error in another field (e.g. `isErrored`, or through a union object), but then the frontend still needs to check on that and build _some_ kind of response from it.

I think that it's a more viable way forward in this case to not have a catch-all for all cases and actually think **about what fields could error**. right now the two that spring to mind are `seriesCount`, and `dataSeries`. we usually only call `dataSeries` for individual nodes, when erroring the parent wouldn't have as bad implications as the user is still able to see the individual insight view card. so we would just go ahead with the current state of this PR.

cc @chwarwick let me know if you have strongly opposing thoughts, or if you need any clarifications.

**Comment by chwarwick** [comment]
In general my thoughts are that somewhere data that can be null or can error needs to be handled and it's best to do that in manner that leaves the application in the best possible state.  I'm also uncomfortable with relying on the application to only query the api in certain ways when we don't enforce that to be true.

I believe that changing the SeriesCount to nullable is not causing any additional issues in the webapp is because the references to it were removed.  If they hadn't been there would still be additional work to address what should happen when a SeriesCount is present vs when it is null.  

Yes SeriesCount and DataSeries are the 2 fields immediately on InsightView that have the potential error currently.  Could they be nullable? From a schema perspective that seems reasonable to me, because if the insights haven't yet backfilled they don't yet have any Series to count or DataSeries to display.  I'm not opposed to that approach but there would be additional UI work to account for those new possibilities.

To me the larger issue is when a list of items is requested do we want to ensure that if all of the items can't be returned that none of them are.  Sometimes that makes sense, like when requesting SeriesPoints,  because I wouldn't want only some of them. In this case with InsightViews I don't think it makes sense, because each InsightView is independent of the others.

My opposition to fixing this problem this way is that it doesn't address the root problem it pushes it out until the next field is added where we have the same debate. 

The root of the problem being that `listOfItems: [Item!]!` means you get everything or nothing.

It would be great to get some more opinions on this @CristinaBirkel  @coury-clark  

I'm not opposed to this change, but if we make it we should make SeriesCount null if the data hasn't been backfilled yet and leave [38951](https://github.com/sourcegraph/sourcegraph/issues/38951) open because we made it less likely to happen but didn't resolve that  problem.

**Comment by felixfbecker** [comment]
> @felixfbecker I'm confused whether you're suggesting this regardless of the fix, or as part of the fix. could you clarify your intentions with documenting the error cases?

If fields are nullable, the UI needs to make changes to handle those cases and for that, it needs to understand in which cases they will be `null`. The best way to communicate that to the frontend team is to update the documentation (even just a few notes/bullet points in the docblock).

**Comment by felixfbecker** [comment]
> In general my thoughts are that somewhere data that can be null or can error needs to be handled and it's best to do that in manner that leaves the application in the best possible state.

> To me the larger issue is when a list of items is requested do we want to ensure that if all of the items can't be returned that none of them are. Sometimes that makes sense, like when requesting SeriesPoints, because I wouldn't want only some of them. In this case with InsightViews I don't think it makes sense, because each InsightView is independent of the others.

I agree with this.

> it requires changes on the frontend such that we always need to handle a potential null node

@vovakulikov why is that not possible/hard to do?

**Comment by vovakulikov** [comment]
@felixfbecker @chwarwick I may misunderstand this problem, but I thought that to solve this problem about nullable fields and errors in general, we should make all fields inside of the `insightView` entity nullable. My main concern was that, in this case, we should have too many default values on the client to handle this situation. 

I see now that @chwarwick mentioned that there are only two fields we should make nullable to solve this, and in this case, I believe it's OK to make them nullable and have some changes on the frontend. 
>Yes, SeriesCount and DataSeries are the two fields immediately on InsightView that have the potential error currently. 


Originally I think we were confused by this suggestion where we have to handle the null `insightView` value in the `InsightViewConnections` array.
> Moving from nodes: [InsightView!]! to nodes: [InsightView]! would mean that an error in any 1 InsightView wouldn't remove the entire list.

In this case, at least to me, it's unclear how we should resolve the null value on the frontend and create an errored insight card, for example, for the dashboard page. @leonore had a prototype where we have all needed and properly resolved fields inside of the `errors` GQL response field, but it seems a bit off and also makes the client parse logic complex (and we're trying to avoid any custom logic around GQL BE types on the frontend because having FE types is a bad decision)

For the record. We also considered having something like this (data in error pattern)
```gql

seriesCount: {
   ...on SuccessSeriesCount {
      value
   }
   
   ... on ErroredSeriesCount {
     __typename
     reason
     errorMesage
     code 
   }
}
```

But I'm not sure about this as well because if we need this pattern for many fields, then all parse logic on the client becomes complex really quick.

**Comment by felixfbecker** [comment]
> In this case, at least to me, it's unclear how we should resolve the null value on the frontend and create an errored insight card, for example, for the dashboard page.

@vovakulikov isn't showing nothing but an error in place of the card still better than showing an error for the entire list? In the case of the entire list being an error, the user can't fix the insight either. This way at least they can still interact with other insights.

I think we should think of these as _unexpected_ errors, e.g. things the user may not even be able to fix (e.g. what if getting data for that insight just timed out?). Any known case we shouldn't express in some generic error struct with `errorMessage: String` etc, but rather as part of the data (e.g. `isStillProcessing: true` on the insight, a union type, etc).

Do we have any _concrete_ case we're actually handling here? Or is this just general protection?

**Comment by vovakulikov** [comment]
@felixfbecker And I see points to make `insightView` inside connection type nullable, but a few thoughts below. 

I'm sorry it's probably overthinking from my side, but what do you propose here?
>I think we should think of these as unexpected errors, e.g. things the user may not even be able to fix (e.g. what if getting data for that insight just timed out?). Any known case we shouldn't express in some generic error struct with errorMessage: String etc, but rather as part of the data (e.g. isStillProcessing: true on the insight, a union type, etc).

If I understood this right, the backend works in a way (which is expected by GQL spec behaviour) that if some unexpected error appears in the field's resolver, we return a null value for this field. If this field is non-nullable in the schema, then GQL propagates this error on the parent level, and this makes the parent null if it's nullable (if it's not, GQL will keep propagating this error to the next parent)  

So in our example, with `nodes: [InsightView]!` we will get something like this if we get an error for the non-nullable `seriesCount`
```gql
nodes : [
   { /* first insight fields */ },
   null,
   { /* third insight fields */ }
]
errors: [
  { path: [...], errorMessage: "", extensions: '../' }
]
```

In this case, it's complicated on the frontend to resolve any useful information to show the insight card with error inside but with some partially resolved information (like we still want to show insight with title, still want to have its id, etc.). 

So the bottom line here and what we want to have on the frontend is
- Have an error as close as possible to the data where this error happened
- Have a direct message field about this error
- Still have partially resolved data about the entity where the error happened, not just the null value.

So I'm probably thinking about this structure, but I'm not sure how scalable this approach is
```gql
insightView {
  ... on SuccessResolvedInsight {
      id
      title, 
     // any other fields 
  }
  ... on Error {
    id
    title
    error { reason, ... }
  }
}
```

Probably just making all error-possible fields nullable (seriesCount and DataSeries) is a better and cheaper option here.

**Comment by leonore** [comment]
**tl;dr** I don't know whether this conversation is getting too big for a PR and would be better taken elsewhere, but in the spirit of focusing our efforts on aggregating search results for 4.0, I'd be inclined to fix this for individual fields and eventually revisit this later. I acknowledge we keep pushing off a lot of graphQL API complexity and this is related to our desired API refactor, and is not ideal.

---

> I believe that changing the SeriesCount to nullable is not causing any additional issues in the webapp is because the references to it were removed. If they hadn't been there would still be additional work to address what should happen when a SeriesCount is present vs when it is null.

@chwarwick I tested this on a commit were `SeriesCount` was still being called, as the test plan of this PR shows, and observed it working correctly.

> Yes SeriesCount and DataSeries are the 2 fields immediately on InsightView that have the potential error currently. Could they be nullable? From a schema perspective that seems reasonable to me, because if the insights haven't yet backfilled they don't yet have any Series to count or DataSeries to display. I'm not opposed to that approach but there would be additional UI work to account for those new possibilities.

@chwarwick As per my point above / this PR shows I don't think there would be additional UI work to account for that, because we error and do not attempt to create a classic insight view card if apollo catches an error.

> I see now that @chwarwick mentioned that there are only two fields we should make nullable to solve this, and in this case, I believe it's OK to make them nullable and have some changes on the frontend.

@vovakulikov so this is the approach I initially suggested and I think that Felix and Chris are kind of opposed to due to it somewhat pushing this issue to the next field we add, if relevant.  

> isn't showing nothing but an error in place of the card still better than showing an error for the entire list? In the case of the entire list being an error, the user can't fix the insight either. This way at least they can still interact with other insights.

@felixfbecker yes it is, and that's what this PR achieves another way. making `InsightView` nullable does achieve the first step of returning both data and errors but makes this actually really hard to handle on the client, as it changes lots of function signatures to now just expect a potentially undefined `InsightView`. I speak from my experience of this as mainly a backend engineer though, and from my discussions with @vovakulikov. It seems like the code for this on the webapp would be clunky.

**Review by chwarwick — APPROVED** [review]
I'm good with making this field nullable, but lets leave the issue open since this doesn't fully fix it.
